Draft of eGPIO SHIM

magp-nordic opened the pull request.

masz-nordic commented on drivers/gpio/egpio_nrfx.c:
```diff
@@ -0,0 +1,194 @@
+#define GPIOTE_PHANDLE(id) DT_INST_PHANDLE(id, gpiote_instance)
+#define GPIOTE_INST(id)    DT_PROP(GPIOTE_PHANDLE(id), instance)
+
+#define GPIOTE_INSTANCE(id)                                     \
+	COND_CODE_1(DT_INST_NODE_HAS_PROP(id, gpiote_instance), \
+		    (NRFX_GPIOTE_INSTANCE(GPIOTE_INST(id))),    \
+		    ({ .p_reg = NULL }))
+
+/* Device instantiation is done with node labels because 'port_num' is
+ * the peripheral number by SoC numbering. We therefore cannot use
+ * DT_INST APIs here without wider changes.
+ */
+
+#define GPIOTE_CHECK(id)						       \
+	COND_CODE_1(DT_INST_NODE_HAS_PROP(id, gpiote_instance),		       \
+		(BUILD_ASSERT(DT_NODE_HAS_STATUS(GPIOTE_PHANDLE(id), okay),    \
+			"Please enable GPIOTE instance for used GPIO port!")), \
+		())
```
I don't think this driver should use GPIOTE nodes.
The only part that should be allocated for it are the pins.
Perhaps a new binding is needed?

magp-nordic committed:
manifest: update hal_nordic version

Update hal_nordic with headers for eGPIO.

Signed-off-by: Magdalena Pastula <magdalena.pastula@nordicsemi.no>

masz-nordic commented on drivers/gpio/Kconfig.nrfx:
```diff
@@ -22,3 +22,18 @@ config GPIO_NRFX_INTERRUPT
 	  significantly reduce memory footprint in case of application that does
 	  not need GPIO interrupts.
+
+choice GPIO_NRFX_TYPE
+	prompt "nRF GPIO driver type"
+	depends on GPIO_NRFX
+	default NRFX_GPIO
+	help
+	  Type of GPIO driver to use.
+
+	config NRFX_GPIO
+		bool "Stardard GPIO"
+
+	config NRFE_GPIO
+		bool "Emulated GPIO"
```
It should be possible to use both at the same time

masz-nordic commented on modules/hal_nordic/nrfx/CMakeLists.txt:
```diff
@@ -19,6 +19,8 @@ zephyr_include_directories(${INC_DIR})
 zephyr_include_directories(${MDK_DIR})
 zephyr_include_directories(.)
 
+zephyr_include_directories_ifdef(CONFIG_NRFE_GPIO ${ZEPHYR_CURRENT_MODULE_DIR}/nrfe)
```
A more generic Kconfig should be used here.

magp-nordic commented:
Blinky sample builds without errors, rebased on https://github.com/zephyrproject-rtos/zephyr/pull/73024
Only the last three commits are valid

masz-nordic commented:
Can you rebase `main` as well? Webpage is lagging for me with this many changes.

magp-nordic committed:
dts: bindings: add new binding for eGPIO

Add binding for emulated GPIO.

Signed-off-by: Magdalena Pastula <magdalena.pastula@nordicsemi.no>

magp-nordic committed:
modules: hal_nordic: add option for emulated peripherals drivers

Add Kconfig option for emulated peripherals drivers support.

Signed-off-by: Magdalena Pastula <magdalena.pastula@nordicsemi.no>

magp-nordic committed:
soc: nordic: nrf54l: add support for NRFE

Add support for emulated peripherals on nRF54L15 APP.

Signed-off-by: Magdalena Pastula <magdalena.pastula@nordicsemi.no>

magp-nordic committed:
drivers: gpio: add eGPIO driver

Add driver for emulated GPIO.

Signed-off-by: Magdalena Pastula <magdalena.pastula@nordicsemi.no>

magp-nordic committed:
samples: basic: blinky: modify to use eGPIO

Modify blinky sample to use emulated GPIO.

Signed-off-by: Magdalena Pastula <magdalena.pastula@nordicsemi.no>

masz-nordic commented on drivers/gpio/CMakeLists.txt:
```diff
@@ -57,6 +57,7 @@ zephyr_library_sources_ifdef(CONFIG_GPIO_NEORV32    gpio_neorv32.c)
 zephyr_library_sources_ifdef(CONFIG_GPIO_NPCX       gpio_npcx.c)
 zephyr_library_sources_ifdef(CONFIG_GPIO_NPM1300    gpio_npm1300.c)
 zephyr_library_sources_ifdef(CONFIG_GPIO_NPM6001    gpio_npm6001.c)
+zephyr_library_sources_ifdef(CONFIG_GPIO_NRFE       egpio_nrfx.c)
```
```suggestion
zephyr_library_sources_ifdef(CONFIG_GPIO_NRFE       gpio_nrfe.c)
```

masz-nordic commented on samples/basic/blinky/prj.conf:
```diff
@@ -1 +1,7 @@
-CONFIG_GPIO=y
+# CONFIG_GPIO=y
```
`CONFIG_GPIO` should be enabled, only `CONFIG_GPIO_NRFX` should be explicitly disabled here.

masz-nordic commented on drivers/gpio/egpio_nrfx.c:
```diff
@@ -0,0 +1,182 @@
+struct ipc_ept ep;
+
+struct gpio_nrfx_data {
+	/* gpio_driver_data needs to be first */
```
Why?

magp-nordic replied:
This comment was copied from GPIO SHIM. I guess it is a Zephyr requirement, that data structures should have a common part that is always first so that the Zephyr layer could access it.

masz-nordic commented on drivers/gpio/egpio_nrfx.c:
```diff
@@ -0,0 +1,182 @@
+	.port_toggle_bits = gpio_nrfx_port_toggle_bits,
+};
+
+#define GPIOTE_PHANDLE(id) DT_INST_PHANDLE(id, gpiote_instance)
+#define GPIOTE_INST(id)    DT_PROP(GPIOTE_PHANDLE(id), instance)
+
+#define GPIOTE_INSTANCE(id)                                     \
+	COND_CODE_1(DT_INST_NODE_HAS_PROP(id, gpiote_instance), \
+		    (NRFX_GPIOTE_INSTANCE(GPIOTE_INST(id))),    \
+		    ({ .p_reg = NULL }))
```
Could this be deleted?

masz-nordic commented on drivers/gpio/egpio_nrfx.c:
```diff
@@ -0,0 +1,182 @@
+
+struct ipc_ept ep;
+
+struct gpio_nrfx_data {
```
```suggestion
struct gpio_nrfe_data {
```
Same in other instances throughout the file.

masz-nordic commented on drivers/gpio/egpio_nrfx.c:
```diff
@@ -0,0 +1,182 @@
+		    ({ .p_reg = NULL }))
+
+
+#define GPIO_NRF_DEVICE(id)						\
```
```suggestion
#define GPIO_NRFE_DEVICE(id)						\
```

masz-nordic reviewed (commented):
Looks good, a few minor comments.
Could you run code formatter to get rid of the compliance fails?

masz-nordic commented on drivers/gpio/gpio_nrfe.c:
```diff
@@ -0,0 +1,158 @@
+
+#define DT_DRV_COMPAT nordic_nrfe_gpio
+
+#include <nrfx_gpiote.h>
+#include <string.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/dt-bindings/gpio/nordic-nrf-gpio.h>
```
I don't think all those headers are necessary.

masz-nordic commented on drivers/gpio/gpio_nrfe.c:
```diff
@@ -0,0 +1,158 @@
+	static struct gpio_nrfe_data gpio_nrfe_p##id##_data;                                       \
+                                                                                                   \
+	DEVICE_DT_INST_DEFINE(id, gpio_nrfe_init, NULL, &gpio_nrfe_p##id##_data,                   \
+			      &gpio_nrfe_p##id##_cfg, POST_KERNEL, CONFIG_GPIO_INIT_PRIORITY,      \
```
Is is always true that GPIO init priority is lower than ipc_service init priority?
Otherwise, wouldn't that cause some issues?

magp-nordic replied:
I added another commit that changes the default init priority when eGPIO is enabled, so that it was initialized after IPC.

masz-nordic commented on modules/hal_nordic/nrfe/Kconfig:
```diff
@@ -0,0 +1,11 @@
+	bool "nRF Emulated Peripherals support"
+	depends on HAS_NRFE
+	help
+	  This option enables the nRF Emulated Peripherals drivers.
```
```suggestion
	  This option enables the nRF Emulated Peripherals.
```

magp-nordic committed:
dts: arm: nordic: nRF54L15: add egpio node

Add node for emulated gpio in nRF54L15 APP dts.

Signed-off-by: Magdalena Pastula <magdalena.pastula@nordicsemi.no>

magp-nordic committed:
soc: nordic: nrf54l: change init priority for eGPIO

Change default init priority for emulated GPIO, so that it is
initialized after IPC.

Signed-off-by: Magdalena Pastula <magdalena.pastula@nordicsemi.no>

magp-nordic commented:
Moved to: https://github.com/masz-nordic/zephyr/pull/1
